tests/ga/test_update.py

```diff
@@ -1680,15 +1683,85 @@ def _mock_popen(cmd, *args, **kwargs):
+                    mock_iptables.set_command(osutil.get_firewall_drop_command(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination), exit_code=0)
+                    # non root tcp iptable rule is absent
+                    mock_iptables.set_command(osutil.get_accept_tcp_rule(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination), exit_code=1)
+                    update_handler.run(debug=True)
+
+                    drop_check_command = TestOSUtil._command_to_string(osutil.get_firewall_drop_command(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination))
+                    accept_tcp_check_rule = TestOSUtil._command_to_string(osutil.get_accept_tcp_rule(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination))
+                    accept_tcp_insert_rule = TestOSUtil._command_to_string(osutil.get_accept_tcp_rule(mock_iptables.wait, AddFirewallRules.INSERT_COMMAND, mock_iptables.destination))
+
+                    # Filtering the mock iptable command calls with only the once related to this test.
+                    filtered_mock_iptable_calls = [cmd for cmd in mock_iptables.command_calls if cmd in [drop_check_command, accept_tcp_check_rule, accept_tcp_insert_rule]]
+
+                    self.assertEqual(len(filtered_mock_iptable_calls), 3, "Incorrect number of calls to iptables: [{0}]".format(mock_iptables.command_calls))
+                    self.assertEqual(filtered_mock_iptable_calls[0], drop_check_command,
+                                     "The first command should check the drop rule")
+                    self.assertEqual(filtered_mock_iptable_calls[1], accept_tcp_check_rule,
+                                     "The second command should check the accept rule")
+                    self.assertEqual(filtered_mock_iptable_calls[2], accept_tcp_insert_rule,
+                                     "The third command should add the accept rule")
+
+    def test_it_should_not_set_dns_tcp_iptable_if_drop_unavailable(self):
+
+        with TestOSUtil._mock_iptables() as mock_iptables:
+            with _get_update_handler(test_data=DATA_FILE) as (update_handler, _):
+                with patch.object(osutil, '_enable_firewall', True):
+                    # drop rule is not available
+                    mock_iptables.set_command(osutil.get_firewall_drop_command(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination), exit_code=1)
+
+                    update_handler.run(debug=True)
+
+                    drop_check_command = TestOSUtil._command_to_string(osutil.get_firewall_drop_command(mock_iptables.wait, AddFirewallRules.CHECK_COMMAND, mock_iptables.destination))
+
+                    # Filtering the mock iptable command calls with only the once related to this test.
+                    filtered_mock_iptable_calls = [cmd for cmd in mock_iptables.command_calls if cmd in [drop_check_command]]
```

Comment:
This will only filter out the drop calls but if the APPEND/CHECK calls for accept_dns rules are there, this test would still succeed. Along with checking if the DROP rule was called, also check if the others were not called to have a more complete test.

Reply:
I agree with the suggestion of adding all the applicable rule to 'filtered_mock_iptable_calls` and I addressed it. However, since we are testing the length of `filtered_mock_iptable_calls` in each test and then checking the order and the particular rule that is called, I don't think it's necessary to check if a rule is not called. If in case a rule is called unintended the length check will fail, if not checks where we test the order will fail. Thoughts on this ?

Reply:
/?how will the length check work here? Basically consider a buggy scenario where we execute 4 firewall commands. When this code is executed (`                    filtered_mock_iptable_calls = [cmd for cmd in mock_iptables.command_calls if cmd in [drop_check_command]]
`), the value of `filtered_mock_iptable_calls` will only contain one value because you're only checking the existence of `drop_check_command`, it wont contain the other 3 firewall calls. In this case, the `len(filtered_mock_iptable_calls)` will always be 1 irrespective of how many firewall calls there are. Is that understanding wrong? Am I missing something else here?